README.md
# Webpack+React集成环境框架

[![Build Status](https://travis-ci.org/greengerong/qing.png?branch=master)](https://travis-ci.org/greengerong/qing)

一个完善的Koa+Webpack+React集成开发环境,实现了多页面应用入口,并附加很多实用组件

## 框架优点

- 支持多页面类型的React应用开发
- 实用Koa服务器进行构建,扩展性更强
- 分为开发环境(Dev)和上线(Prod)环境,可以自由选择并进行开发
- 支持React热加载,可以实现CSS及JS的热加载效果
- 附加Postcss工具,可实现Sass语法开发,自动填充CSS3兼容方案及CSS雪碧图的自动化构建
- 使用了CSS-Module技术,实现了样式的复用性和解决了样式污染**(在dist/css/common和dist/css/components中的样式不会启用CSS-Module**

##使用的技术

- [x] [Webpack](https://webpack.github.io)
- [x] [React](https://facebook.github.io/react/)
- [x] [Babel](https://babeljs.io/)
- [x] [Koa](https://github.com/koajs/koa)
- [x] [PostCSS](https://github.com/postcss/postcss)
- [x] [precss](https://github.com/jonathantneal/precss)
- [x] [CSS modules](https://github.com/outpunk/postcss-modules)
- [x] [autoprefixer](https://github.com/postcss/autoprefixer)
- [x] [webpack-dashboard](https://github.com/FormidableLabs/webpack-dashboard)

## 使用方法
**1.安装所需NodeJS模块**

进入你该项目文件夹,并在CMD/终端中执行命令安装Node依赖包,这里有一个需要注意的问题,由于webpack中file-loader的图片处理路径无法满足CSS背景图片的路径变化,需要将根目录下的node_bak中的file-loader文件夹替换node_modules中的file-loader文件夹
```
npm install
```
![IMG](./README/3.png)

**2.创建入口及对应js文件**

在src开发目录中创建HMTL页面,并在src/dist/js中创建一个同名js即可

![IMG](./README/1.png)

**3.开启热加载开发模式**

进入你该项目文件夹,并在CMD/终端中执行命令开启开发模式,这时候会显示一个控制台,能够看到项目对应的反馈信息及问题

```
npm start
```
**4.项目完成后进行打包压缩**

进入你该项目文件夹,并在CMD/终端中执行命令执行打包压缩,会在根目录生成一个assets的文件夹,文件夹内为压缩后的项目代码

```
npm run build 
```

## 其他框架特性内容介绍

**1.PostCss工具**

在框架中使用了PostCss的部分工具,主要为

- [Precss](https://jonathantneal.github.io/precss/) 可以实现类似sass语法功能,所以该框架支持类似Sass语法的编写,例如:
```css
/* before */

@for $i from 1 to 3 {
	.b-$i { width: $(i)px; }
}

/* after */

.b-1 {
	width: 1px
}
.b-2 {
	width: 2px
}
.b-3 {
	width: 3px
}
```
- [postcss-easysprites](https://github.com/glebmachine/postcss-easysprites) 可以实现图表的自动拼接,例如:

![IMG](./README/fork.png)![IMG](./README/plus.png)![IMG](./README/github.png)![IMG](./README/plus.png)![IMG](./README/twitter.png)![IMG](./README/equals.png)

![IMG](./README/spritesheet.png)
```css
/* before */

.arrowNext {
  background-image: url('dist/img/arrow-next.png#elements');
}
.arrowNext:hover {
  background-image: url('dist/img/arrow-next_hover.png#elements');
}
.arrowPrev {
  background-image: url('dist/img/arrow-prev.png#elements2');
}
.arrowPrev:hover {
  background-image: url('dist/img/arrow-prev_hover.png#elements2');
}

/* after */

.arrowNext { 
  background-image: url(dist/img/elements.png); 
  background-position: 0 0;
}
.arrowNext:hover { 
  background-image: url(dist/img/elements.png); 
  background-position: -28px 0;
}
.arrowPrev { 
  background-image: url(dist/img/elements2.png); 
  background-position: 0 0;
}
.arrowPrev:hover { 
  background-image: url(dist/img/elements2.png); 
  background-position: -28px 0;
}
```
## 更新记录

- [CHANGELOG](https://github.com/aemoe/webpack-react-framework/blob/master/CHANGELOG.md)

# License 

MIT



CHANGELOG.md
## 1.0.3 (2016-10-17)

Update:

  - Add a react-router demo(router.html)

  - 增加了一个react-router实例(router.html)


## 1.0.2 (2016-09-13)

Update:

  - Add a listview demo,and update Readme

  - 增加了一个列表实例,并优化补充了Readme


## 1.0.1 (2016-09-05)

Update:

  - Edit .babelrc file config.
  In order to the system suppose class-properties and ... that it added babel-preset-stage-2 and babel-plugin-transform-class-properties.

  - 修改了.babelrc的配置
  为了让系统支持ES7的扩展方法,增加2个bable的插件




config/devServer.js
//加载Node的Path模块
const path = require('path');
//加载webpack模块
const webpack = require('webpack');
//加载webpack 热加载服务器模块
const webpackDevMiddleware = require('koa-webpack-dev-middleware');
const webpackHotMiddleware = require('koa-webpack-hot-middleware');
//加载webpack-dashboard监控视窗模块
const Dashboard = require('webpack-dashboard');
const DashboardPlugin = require('webpack-dashboard/plugin');
//加载webpack配置文件
const config = require('./webpack.config.dev');
//加载koa服务器模块
const koa = require('koa');
const fetch = require('isomorphic-fetch')

//配置及初始化Koa服务器
var creatServer = () => {
    //创建koa服务器应用
    let app = koa();
    //初始化webpack应用
    let compiler = webpack(config);
    //初始化webpack-dashboard应用
    let dashboard = new Dashboard();
    //调用webpack-dashboard应用
    compiler.apply(new DashboardPlugin(dashboard.setData));
    //调用webpack热加载模块及对应参数
    app.use(webpackDevMiddleware(compiler, {
        quiet: true,
        noInfo: true,
        publicPath: config.output.publicPath,
        hot: true,
        lazy: false,
        historyApiFallback: true
    }));
    app.use(webpackHotMiddleware(compiler), {
        log: () => {}
    });
    //调用koa静态文件
    app.use(require('koa-static')(path.join(process.cwd(), '/assets')));
    //调用koa列出开发目录
    app.use(require('koa-serve-index')(path.join(process.cwd(), '/src')));
    // 代理
    app.use(function *(next){
        // console.log(this.request)
        if(this.request.method === "GET") {
            // 判断是否是 api 请求
            // let pre = path.parse(this.request.url)[0]
            // console.log(pre)
            if(this.request.url.indexOf('api') > -1) {
                // console.log(this.request.url.slice(3))
                try {
                    let res = yield fetch('http://web.juhe.cn:8080'+ this.request.url.slice(4)) 
                    // console.log(res)
                    let json = yield res.json()
                    if(this.request)
                    this.body = json
                    next
                }catch(err) {
                    this.body = 'not found'
                    next
                }
            }
            next
        }
        next
    });
    //调用开启5000端口用来测试和开发
    app.listen(5000, function(err) {
        if (err) { console.log(err); }
        console.log('Listening at localhost:5000');
    });
};


//调用创建koa服务器方法
creatServer();


config/webpack.config.prod.js
'use strict';
//加载Node的Path模块
var path = require('path');
//加载Node的fs模块
var fs = require('fs');
//加载webpack模块
var webpack = require('webpack');
//srcDir为当前开发目录(/src)
var srcDir = path.resolve(process.cwd(), 'src');
//srcDir为当前建立目录(/assets)
var assetsDir = path.resolve(process.cwd(), 'assets');
//加载自动化css独立加载插件
var ExtractTextPlugin = require('extract-text-webpack-plugin');
//加载自动化HTML自动化编译插件
var HtmlWebpackPlugin = require('html-webpack-plugin');
//加载JS模块压缩编译插件
var UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
//加载公用组件插件
var CommonsChunkPlugin = webpack.optimize.CommonsChunkPlugin;
//设置需要排除单独打包的插件
var singleModule = ['react', 'react-dom', 'jquery', 'Raphael'];
//postcss辅助插件
var postcssImport = require('postcss-import');
//排除的页面入口js
var jsExtract = [];

//加载webpack目录参数配置
var config = {
    devtool: 'source-map',
    entry: getEntry(),
    output: {
        path: path.join(process.cwd(), 'assets'),
        filename: 'dist/js/[name].js',
        publicPath: ''
    },
    plugins: [
        //排除css压缩加载在页面
        new ExtractTextPlugin('dist/css/[name].css'),
        //合并额外的js包
        new CommonsChunkPlugin('lib', './dist/js/lib.js', jsExtract),
        //new webpack.optimize.CommonsChunkPlugin('lib', './dist/js/lib.js'),
        new webpack.optimize.UglifyJsPlugin({
            compressor: {
                warnings: false
            }
        }),
        new webpack.ProvidePlugin({
            $: 'jquery'
        })
    ],
    module: {
        //加载器配置
        loaders: [{
            test: /\.css$/,
            exclude: [path.resolve(process.cwd(), 'src/dist/css/common'), path.resolve(process.cwd(), 'src/dist/css/components')],
            loader: ExtractTextPlugin.extract('style', 'css?modules&localIdentName=[name]__[local]___[hash:base64:5]!postcss-loader')
        }, {
            test: /\.css$/,
            include: [path.resolve(process.cwd(), 'src/dist/css/common'), path.resolve(process.cwd(), 'src/dist/css/components')],
            loader: ExtractTextPlugin.extract('style', 'css!postcss-loader')
        }, {
            test: /\.js$/,
            loaders: ['babel'],
            exclude: /node_modules/, // 匹配不希望处理文件的路径
            include: path.join(process.cwd(), 'src')
        }, {
            test: /\.(png|jpeg|jpg|gif)$/,
            loader: 'file?name=../img/[name].[ext]&to=dist/img/[name].[ext]'
        }, {
            test: /\.(woff|eot|ttf)$/i,
            loader: 'url?limit=10000&name=dist/fonts/[name].[ext]'
        }, {
            test: /\.json$/,
            loader: 'json'
        }]
    },
    postcss: function(webpack) {
        return [
            postcssImport({
                addDependencyTo: webpack
            }),
            require('postcss-display-inline-block'),
            require('autoprefixer'),
            require('precss'),
            require('postcss-easysprites')({
                imagePath: '../img',
                spritePath: './assets/dist/img'
            })
        ];
    }
};

//设置入口文件
function getEntry() {
    var jsDir = path.resolve(srcDir, 'dist/js');
    var names = fs.readdirSync(jsDir);
    var map = {};
    names.forEach(function(name) {
        var m = name.match(/(.+)\.js$/);
        var entry = m ? m[1] : '';
        var entryPath = entry ? path.resolve(jsDir, name) : '';
        if (entry) {
            jsExtract.push(name.substring(0, name.length - 3));
            map[entry] = entryPath;
        }
    });
    //自定义额外加载包,不会合并在页面对应js中
    map['lib'] = singleModule;
    return map;
}

var pages = fs.readdirSync(srcDir);
pages.forEach(function(filename) {
    var m = filename.match(/(.+)\.html$/);
    if (m) {
        var conf = {
            template: path.resolve(srcDir, filename),
            inject: true, //允许插件修改哪些内容，包括head与body
            hash: true, //为静态资源生成hash值
            minify: { //压缩HTML文件
                removeComments: true, //移除HTML中的注释
                collapseWhitespace: false //删除空白符与换行符
            },
            filename: filename
        };

        if (m[1] in config.entry) {
            conf.chunks = ['vendors', m[1]];
        }

        config.plugins.push(new HtmlWebpackPlugin(conf));
    }
});

module.exports = config;


node_bak/file-loader/index.js
/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
var loaderUtils = require("loader-utils");

module.exports = function(content) {
    this.cacheable && this.cacheable();
    if (!this.emitFile) throw new Error("emitFile is required from module system");
    var query = loaderUtils.parseQuery(this.query);
    var url = loaderUtils.interpolateName(this, query.name || "[hash].[ext]", {
        context: query.context || this.options.context,
        content: content,
        regExp: query.regExp
    });
    var to = (query.to) && loaderUtils.interpolateName(this, query.to, {
        context: query.context || this.options.context,
        content: content,
        regExp: query.regExp
    });
    this.emitFile((to) ? to : url, content);
    return "module.exports = __webpack_public_path__ + " + JSON.stringify(url) + ";";
}
module.exports.raw = true;

config/webpack.config.dev.js
'use strict';
//加载Node的Path模块
var path = require('path');
//加载Node的fs模块
var fs = require('fs');
//加载webpack模块
var webpack = require('webpack');
//srcDir为当前开发目录(/src)
var srcDir = path.resolve(process.cwd(), 'src');
//srcDir为当前建立目录(/assets)
var assetsDir = path.resolve(process.cwd(), 'assets');
//加载自动化css独立加载插件
var ExtractTextPlugin = require('extract-text-webpack-plugin');
//加载自动化HTML自动化编译插件
var HtmlWebpackPlugin = require('html-webpack-plugin');
//加载JS模块压缩编译插件
var UglifyJsPlugin = webpack.optimize.UglifyJsPlugin;
//加载公用组件插件
var CommonsChunkPlugin = webpack.optimize.CommonsChunkPlugin;
//设置需要排除单独打包的插件
var singleModule = ['react', 'react-dom', 'jquery', 'Raphael'];
//排除的页面入口js
var jsExtract = [];

//获取本级IP 以方便手机端测试使用
function getIPAdress() {
    var interfaces = require('os').networkInterfaces();
    for (var devName in interfaces) {
        var iface = interfaces[devName];
        for (var i = 0; i < iface.length; i++) {
            var alias = iface[i];
            if (alias.family === 'IPv4' && alias.address !== '127.0.0.1' && !alias.internal) {
                return alias.address;
            }
        }
    }
}

//加载webpack目录参数配置
var config = {
    devtool: 'cheap-module-eval-source-map',
    entry: getEntry(),
    output: {
        path: path.join(process.cwd(), 'assets'),
        filename: 'dist/js/[name].js',
        publicPath: 'http://' + getIPAdress() + ':5000/'
    },
    plugins: [
        //排除css压缩加载在页面
        new ExtractTextPlugin('dist/css/[name].css'),
        //合并额外的js包
        new CommonsChunkPlugin('lib', './dist/js/lib.js', jsExtract),
        new webpack.HotModuleReplacementPlugin(),
        new webpack.NoErrorsPlugin(),
        new webpack.ProvidePlugin({
            $: 'jquery'
        })
    ],
    module: {
        //加载器配置
        loaders: [{
            test: /\.css$/,
            exclude: [path.resolve(process.cwd(), 'src/dist/css/common'), path.resolve(process.cwd(), 'src/dist/css/components')],
            loaders: [
                'style-loader',
                'css-loader?modules&localIdentName=[name]__[local]___[hash:base64:5]&sourceMap&importLoaders=1',
                'postcss-loader?sourceMap=true'
            ]
        }, {
            test: /\.css$/,
            include: [path.resolve(process.cwd(), 'src/dist/css/common'), path.resolve(process.cwd(), 'src/dist/css/components')],
            loaders: [
                'style-loader',
                'css-loader?sourceMap&importLoaders=1',
                'postcss-loader?sourceMap=true'
            ]
        }, {
            test: /\.js$/,
            loaders: ['react-hot', 'babel'],
            exclude: /node_modules/, // 匹配不希望处理文件的路径
            include: path.join(process.cwd(), 'src')
        }, {
            test: /\.(png|jpeg|jpg|gif)$/,
            loader: 'file?name=dist/img/[name].[ext]'
        }, {
            test: /\.(woff|eot|ttf)$/i,
            loader: 'url?limit=10000&name=dist/fonts/[name].[ext]'
        }, {
            test: /\.json$/,
            loader: 'json'
        }]
    },
    postcss: function(webpack) {
        return {
            plugins: [
                require('postcss-import')({
                    addDependencyTo: webpack
                }),
                require('postcss-display-inline-block'),
                require('autoprefixer'),
                require('precss'),
                require('postcss-easysprites')({
                    imagePath: '../img',
                    spritePath: './assets/dist/img'
                })
            ]
        };
    }
};

//设置入口文件
function getEntry() {
    var jsDir = path.resolve(srcDir, 'dist/js');
    var names = fs.readdirSync(jsDir);
    var map = {};
    names.forEach(function(name) {
        var m = name.match(/(.+)\.js$/);
        var entry = m ? m[1] : '';
        var entryPath = entry ? path.resolve(jsDir, name) : '';
        var entryArr = [];
        entryArr.push(entryPath);
        entryArr.push('eventsource-polyfill');
        entryArr.push('webpack-hot-middleware/client');
        jsExtract.push(name);
        if (entry) {
            jsExtract.push(name.substring(0, name.length - 3));
            map[entry] = entryArr;
        }
    });
    //自定义额外加载包,不会合并在页面对应js中
    map['lib'] = singleModule;
    return map;
}

var pages = fs.readdirSync(srcDir);
pages.forEach(function(filename) {
    var m = filename.match(/(.+)\.html$/);
    if (m) {
        var conf = {
            template: path.resolve(srcDir, filename),
            inject: true, //允许插件修改哪些内容，包括head与body
            hash: true, //为静态资源生成hash值
            minify: { //压缩HTML文件
                removeComments: true, //移除HTML中的注释
                collapseWhitespace: false //删除空白符与换行符
            },
            filename: filename
        };

        if (m[1] in config.entry) {
            conf.chunks = ['vendors', m[1]];
        }

        config.plugins.push(new HtmlWebpackPlugin(conf));
    }
});

module.exports = config;
